### ISSUE TITLE
How to choose dtype?

### ISSUE THREAD

**Issue by hyansuper** [spatialaudio/python-sounddevice#240]
I am not a expert on file format, I am trying to stream raw data to / from a raspberry pi, using soundfile.RawOutputStream / RawInputStream, sound quality is not priority concern, I assume `uint8` and `int8` dtype is most resource-saving on network, or maybe not so mean, I can use other dtypes.

Judging by name, I assume `uint8` and `int8`, `float32` and `int32` takes up same amount of space? what's their difference? Can I set dtype as I want or is it specific to device hardware?

**Comment by mgeier** [comment]
> I assume `uint8` and `int8` dtype is most resource-saving on network

Yes, as the name suggests, each audio sample will need 8 bits (i.e. 1 byte).
But the sound quality will be very poor. Speech will sound kinda bad, music will sound very bad.

If you want to save network bandwidth, it's probably better to use compression?

For just twice the space, you could use `int16`, which will sound much better.

> Can I set dtype as I want or is it specific to device hardware?

It might be, I'm not sure.

You can use [check_input_settings()](https://python-sounddevice.readthedocs.io/en/0.3.15/api/checking-hardware.html#sounddevice.check_input_settings) to see whether a given format is supported.

`uint8` (the `u` means "unsigned") is a very un-common audio format. It doesn't oscillate around 0, so this is strange. It has a range from `0` to `255`.
But if it works on your hardware, you can of course use it if you want.

And yes, `int8` takes up the same amount of space. But it is "signed", i.e. it oscillates around 0 and its values have a range from `-128` to `127`.

See also https://python-sounddevice.readthedocs.io/en/0.3.15/api/module-defaults.html#sounddevice.default.dtype

> `float32` and `int32` takes up same amount of space?

Yes.

> what's their difference?

One is "floating point" and the other is just integers. `float32` has a range from `-1.0` to `1.0`, while `int32` has a range from `-2147483648` (`-2**31`) to `2147483647` (`2**31 - 1`).

In my experience, `int32` is very rarely used.
`float32` is often used for internal calculations within a program, and also for interchange between programs.

In the end, it depends on your requirements. If you use more than one byte per sample (or more than one channel), you'll also have to think about how you send the data over network in a way that the other side knows where the stream of data starts. You also have to take care of the order of bytes (i.e. the "endianness"), otherwise you might receive some horrible noise on the other side.
